buildSrc/subprojects/performance/src/main/groovy/org/gradle/testing/PerformanceTest.groovy

```diff
@@ -14,109 +14,104 @@
  * limitations under the License.
  */
 
-package org.gradle.testing;
+package org.gradle.testing
 
-import org.gradle.api.internal.tasks.options.Option;
-import org.gradle.api.tasks.Input;
-import org.gradle.api.tasks.Optional;
-import org.gradle.api.tasks.OutputDirectory;
-import org.gradle.gradlebuild.test.integrationtests.DistributionTest;
+import groovy.transform.CompileStatic
+import org.gradle.api.internal.tasks.options.Option
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.gradlebuild.test.integrationtests.DistributionTest
 
-import javax.annotation.Nullable;
-import java.io.File;
+import javax.annotation.Nullable
 
 /**
  * A test that checks execution time and memory consumption.
  */
-public class PerformanceTest extends DistributionTest {
+@CacheableTask
+@CompileStatic
+class PerformanceTest extends DistributionTest {
+    @Nullable
+    String baselines
 
-    private String baselines;
-    private String warmups;
-    private String runs;
-    private String checks;
-    private String channel;
-    private File debugArtifactsDirectory = new File(getProject().getBuildDir(), getName());
+    @Nullable
+    String warmups
 
-    @Option(option = "scenarios", description = "A semicolon-separated list of performance test scenario ids to run.")
-    public void setScenarios(String scenarios) {
-        systemProperty("org.gradle.performance.scenarios", scenarios);
```

Comment:
Let's please do the following here:
- Move all the things which influence system properties to an inner `CommandLineArgumentProvider` class
- this class should declare the right inputs/outputs via annotations and construct the system properties command line argument from the fields of the class
- Add an instance of that class to the test task via `jvmArgumentProviders.add()`
- Remove the hack from the plugin and use the system property value as default instead: https://github.com/gradle/gradle/blob/96cf104f70b630fe408cac2172a0868a989da657/buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/plugins/performance/PerformanceTestPlugin.kt#L432-L436

Then we can declare the flamegraph target directory as an output and don't need to pass a relative path around.

Reply:
Also, flamegraph output directory is a subdirectory of `debugArtifactsDirectory`, which is already annotated with `@OutputDirectory` so I don't think it necessary to annotate flamegraph directory again.